pkg/recipes/terraform/config/providers/types.go

```diff
@@ -47,26 +49,57 @@ func GetUCPConfiguredTerraformProviders(ucpConn sdk.Connection, secretProvider *
 
 // GetRecipeProviderConfigs returns the Terraform provider configurations for Terraform providers
 // specified under the RecipeConfig/Terraform/Providers section under environment configuration.
-func GetRecipeProviderConfigs(ctx context.Context, envConfig *recipes.Configuration) map[string][]map[string]any {
+func GetRecipeProviderConfigs(ctx context.Context, envConfig *recipes.Configuration, secrets map[string]map[string]string) (map[string][]map[string]any, error) {
 	providerConfigs := make(map[string][]map[string]any)
 
 	// If the provider is not configured, or has empty configuration, skip this iteration
 	if envConfig != nil && envConfig.RecipeConfig.Terraform.Providers != nil {
 		for provider, config := range envConfig.RecipeConfig.Terraform.Providers {
-			if len(config) > 0 {
-				configList := make([]map[string]any, 0)
-
-				// Retrieve configuration details from 'AdditionalProperties' property and add to the list.
-				for _, configDetails := range config {
-					if configDetails.AdditionalProperties != nil && len(configDetails.AdditionalProperties) > 0 {
-						configList = append(configList, configDetails.AdditionalProperties)
-					}
+			configList := make([]map[string]any, 0)
+
+			for _, configDetails := range config {
+				// Create map for current config
+				currentConfig := make(map[string]any)
+
+				// Retrieve configuration details from 'AdditionalProperties' property and add to currentConfig.
+				if configDetails.AdditionalProperties != nil && len(configDetails.AdditionalProperties) > 0 {
+					currentConfig = configDetails.AdditionalProperties
+				}
+
+				// Extract secrets from configDetails if they are present and update the currentConfig map
+				err := extractSecretsFromConfig(currentConfig, configDetails.Secrets, secrets)
+				if err != nil {
+					return nil, err
+				}
+
+				if len(currentConfig) > 0 {
+					configList = append(configList, currentConfig)
 				}
 
 				providerConfigs[provider] = configList
 			}
 		}
 	}
 
-	return providerConfigs
+	return providerConfigs, nil
+}
+
+// extractSecretsFromConfig extracts secrets for env recipe configuration from the secrets data input and updates the currentConfig map.
+func extractSecretsFromConfig(currentConfig map[string]any, recipeConfigSecrets map[string]datamodel.SecretReference, secrets map[string]map[string]string) error {
+	// Extract secrets from configDetails if they are present
+	for secretName, secretReference := range recipeConfigSecrets {
+
+		// Extract secret value from the secrets data input
+		if secretSource, ok := secrets[secretReference.Source]; ok {
+			if secretValue, ok := secretSource[secretReference.Key]; ok {
+				currentConfig[secretName] = secretValue
```

Comment:
Can this be overridden by any chance? I mean, can `currentConfig[secretName]` be already not empty at this point?

Reply:
This is a good question. The one way that it is possible is if the user puts in the same key in provider configuration and under secrets for a provider configuration. eg bicep input:
```
          azurerm: [
            {
              subscriptionid: 1234
              tenant_id: '745fg88bf-86f1-41af'
              client_id: 'abc123'
              secrets: { 
                client_id: {
                  source: secretStoreConfig.id
                  key: 'secretkey.client_id'
                }
                password: {
                  source: secretStoreConfig.id
                  key: 'secretkey.password'
                }
              }
            },
```
 The current code will have the Secret value override any existing values for the same key in the provider configuration. We can review if this is desired/acceptable behaviour.

Reply:
taking from a (hotly debated) example from the Hasicorp Vault project, seems like they opted to keep the overwriting behavior with the behavior documented:
- https://github.com/hashicorp/vault/issues/182
- https://github.com/hashicorp/vault/commit/2d9b12b85306290b4b58cd764cf0f8fa0d0f5b17

Thus, I think it's ok to proceed with the overwriting behavior.

Reply:
Thanks @willtsai. I have added a test for this case. We'll get @kachawla input on this and then I will document this behaviour in the design doc.